{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08114924-bff1-4ded-8357-f33a3f7e0cc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, time, tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "226c5862-4e39-4457-8e84-d93d1b8de506",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2300000</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5400000</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5400000</td>\n",
       "      <td>7200000</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7200000</td>\n",
       "      <td>9200000</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9200000</td>\n",
       "      <td>12700000</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1         2       3       4\n",
       "0  chr1        0   2300000  p36.33    gneg\n",
       "1  chr1  2300000   5400000  p36.32  gpos25\n",
       "2  chr1  5400000   7200000  p36.31    gneg\n",
       "3  chr1  7200000   9200000  p36.23  gpos25\n",
       "4  chr1  9200000  12700000  p36.22    gneg"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoBand = pd.read_csv('../ref/hg19_cytoBand.txt', sep='\\t', header=None)\n",
    "cytoBand.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81793cc9-d5a8-4dbd-bdd6-6908c6f0d673",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_988886/3348586129.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ukb_variants = pd.read_csv('../blood_coloc/gwas/ukb.variants.tsv.bgz', compression='gzip', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(13791467, 25)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ukb_variants = pd.read_csv('../blood_coloc/gwas/ukb.variants.tsv.bgz', compression='gzip', sep='\\t')\n",
    "ukb_variants.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb8cb561-18d6-47f9-90be-de12179f56ef",
   "metadata": {},
   "source": [
    "# RA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "160ec1a9-4b08-4f2e-a516-94aa951cca12",
   "metadata": {},
   "source": [
    "## GWAS preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "25a663c8-c181-47ff-a285-cf239cd7e8bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/RA-10-28-2021/EUR_all_auto-10-2021.txt.gz', sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "be540ed2-0f4a-43ad-b9d9-e2a74ed9f670",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = pd.read_csv('../ref/reference.1000G.maf.0.005.txt', sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "09c1e357-7914-4a3a-8f92-e35e7a73a21d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>MAF</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs201725126</td>\n",
       "      <td>1</td>\n",
       "      <td>13116</td>\n",
       "      <td>0.186879</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs200579949</td>\n",
       "      <td>1</td>\n",
       "      <td>13118</td>\n",
       "      <td>0.186879</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs75454623</td>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>0.479125</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs199856693</td>\n",
       "      <td>1</td>\n",
       "      <td>14933</td>\n",
       "      <td>0.050696</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs78601809</td>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>0.268390</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP  CHR     BP       MAF A1 A2\n",
       "0  rs201725126    1  13116  0.186879  T  G\n",
       "1  rs200579949    1  13118  0.186879  A  G\n",
       "2   rs75454623    1  14930  0.479125  A  G\n",
       "3  rs199856693    1  14933  0.050696  G  A\n",
       "4   rs78601809    1  15211  0.268390  T  G"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "f253dc11-38f1-465f-b251-37f02447534e",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw[['CHR', 'BP', 'ref', 'alt']] = raw.SNP.str.split('_', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "e2f31ed4-a25f-4346-aaad-71100263277d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pval</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10_26354101_G_T</td>\n",
       "      <td>-0.0186</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.2200</td>\n",
       "      <td>10</td>\n",
       "      <td>26354101</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3_16794282_CTGTT_C</td>\n",
       "      <td>-0.0079</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>0.5711</td>\n",
       "      <td>3</td>\n",
       "      <td>16794282</td>\n",
       "      <td>CTGTT</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13_38855689_T_C</td>\n",
       "      <td>-0.0517</td>\n",
       "      <td>0.0828</td>\n",
       "      <td>0.5325</td>\n",
       "      <td>13</td>\n",
       "      <td>38855689</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4_137301978_T_C</td>\n",
       "      <td>1.3959</td>\n",
       "      <td>0.5448</td>\n",
       "      <td>0.0104</td>\n",
       "      <td>4</td>\n",
       "      <td>137301978</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18_4865341_T_A</td>\n",
       "      <td>0.1558</td>\n",
       "      <td>0.1372</td>\n",
       "      <td>0.2561</td>\n",
       "      <td>18</td>\n",
       "      <td>4865341</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SNP    Beta      SE    Pval CHR         BP    ref alt\n",
       "0     10_26354101_G_T -0.0186  0.0151  0.2200  10   26354101      G   T\n",
       "1  3_16794282_CTGTT_C -0.0079  0.0140  0.5711   3   16794282  CTGTT   C\n",
       "2     13_38855689_T_C -0.0517  0.0828  0.5325  13   38855689      T   C\n",
       "3     4_137301978_T_C  1.3959  0.5448  0.0104   4  137301978      T   C\n",
       "4      18_4865341_T_A  0.1558  0.1372  0.2561  18    4865341      T   A"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "2e80caa3-8b82-4900-9989-4c8f6535d6e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SNP      object\n",
       "Beta    float64\n",
       "SE      float64\n",
       "Pval    float64\n",
       "CHR       int64\n",
       "BP        int64\n",
       "ref      object\n",
       "alt      object\n",
       "dtype: object"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = raw.astype({'CHR': int, 'BP': int})\n",
    "raw.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "35d90974-15e3-4629-be31-db38d55a3831",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.merge(raw, ref, on=['CHR', 'BP'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "4aec2dc7-96e6-4166-b0d2-100efce1df41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((8919821, 12), (13297690, 8), (9667224, 6))"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.shape, raw.shape, ref.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "13b7c252-d18b-4f99-a4f3-4d53edca0c5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_x</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pval</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>SNP_y</th>\n",
       "      <th>MAF</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10_26354101_G_T</td>\n",
       "      <td>-0.0186</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.2200</td>\n",
       "      <td>10</td>\n",
       "      <td>26354101</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>rs10828946</td>\n",
       "      <td>0.438370</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13_38855689_T_C</td>\n",
       "      <td>-0.0517</td>\n",
       "      <td>0.0828</td>\n",
       "      <td>0.5325</td>\n",
       "      <td>13</td>\n",
       "      <td>38855689</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs12017161</td>\n",
       "      <td>0.008946</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9_100158876_T_G</td>\n",
       "      <td>-0.0076</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.6217</td>\n",
       "      <td>9</td>\n",
       "      <td>100158876</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>rs4742689</td>\n",
       "      <td>0.381710</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3_141430083_G_A</td>\n",
       "      <td>-0.0094</td>\n",
       "      <td>0.0795</td>\n",
       "      <td>0.9061</td>\n",
       "      <td>3</td>\n",
       "      <td>141430083</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs16851691</td>\n",
       "      <td>0.009940</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19_38452535_C_T</td>\n",
       "      <td>-0.0128</td>\n",
       "      <td>0.0203</td>\n",
       "      <td>0.5290</td>\n",
       "      <td>19</td>\n",
       "      <td>38452535</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs353416</td>\n",
       "      <td>0.172962</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             SNP_x    Beta      SE    Pval  CHR         BP ref alt  \\\n",
       "0  10_26354101_G_T -0.0186  0.0151  0.2200   10   26354101   G   T   \n",
       "1  13_38855689_T_C -0.0517  0.0828  0.5325   13   38855689   T   C   \n",
       "2  9_100158876_T_G -0.0076  0.0154  0.6217    9  100158876   T   G   \n",
       "3  3_141430083_G_A -0.0094  0.0795  0.9061    3  141430083   G   A   \n",
       "4  19_38452535_C_T -0.0128  0.0203  0.5290   19   38452535   C   T   \n",
       "\n",
       "        SNP_y       MAF A1 A2  \n",
       "0  rs10828946  0.438370  G  T  \n",
       "1  rs12017161  0.008946  T  C  \n",
       "2   rs4742689  0.381710  T  G  \n",
       "3  rs16851691  0.009940  G  A  \n",
       "4    rs353416  0.172962  C  T  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "42c52856-886f-48a6-84d2-5f957fbfa75a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8919817, 12)"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged2 = merged.loc[((merged.ref==merged.A1) & (merged.alt==merged.A2)) | ((merged.ref==merged.A2) & (merged.alt==merged.A1))]\n",
    "merged2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "100af5f4-ad7a-4e31-9dd8-4c9b06b3070b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3686288/282468429.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged3.rename(columns={'SNP_y': 'SNP'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>26354101</td>\n",
       "      <td>rs10828946</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0186</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.2200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>38855689</td>\n",
       "      <td>rs12017161</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0517</td>\n",
       "      <td>0.0828</td>\n",
       "      <td>0.5325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>100158876</td>\n",
       "      <td>rs4742689</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0076</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.6217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>141430083</td>\n",
       "      <td>rs16851691</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0094</td>\n",
       "      <td>0.0795</td>\n",
       "      <td>0.9061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>38452535</td>\n",
       "      <td>rs353416</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0128</td>\n",
       "      <td>0.0203</td>\n",
       "      <td>0.5290</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR         BP         SNP alt ref    Beta      SE    Pval\n",
       "0   10   26354101  rs10828946   T   G -0.0186  0.0151  0.2200\n",
       "1   13   38855689  rs12017161   C   T -0.0517  0.0828  0.5325\n",
       "2    9  100158876   rs4742689   G   T -0.0076  0.0154  0.6217\n",
       "3    3  141430083  rs16851691   A   G -0.0094  0.0795  0.9061\n",
       "4   19   38452535    rs353416   T   C -0.0128  0.0203  0.5290"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged3 = merged2[['CHR', 'BP', 'SNP_y', 'alt', 'ref', 'Beta', 'SE', 'Pval']]\n",
    "merged3.rename(columns={'SNP_y': 'SNP'}, inplace=True)\n",
    "merged3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "d8197cce-02f3-49ef-87e9-e658daa4ef96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>rs202152658</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0588</td>\n",
       "      <td>0.0380</td>\n",
       "      <td>0.1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>751756</td>\n",
       "      <td>rs143225517</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0585</td>\n",
       "      <td>0.0379</td>\n",
       "      <td>0.1230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>rs3094315</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>0.0328</td>\n",
       "      <td>0.1495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.2354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>752894</td>\n",
       "      <td>rs3131971</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0522</td>\n",
       "      <td>0.0357</td>\n",
       "      <td>0.1439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR      BP          SNP alt ref    Beta      SE    Pval\n",
       "0    1  751343  rs202152658   A   T -0.0588  0.0380  0.1220\n",
       "1    1  751756  rs143225517   C   T -0.0585  0.0379  0.1230\n",
       "2    1  752566    rs3094315   A   G  0.0472  0.0328  0.1495\n",
       "3    1  752721    rs3131972   G   A  0.0325  0.0274  0.2354\n",
       "4    1  752894    rs3131971   C   T  0.0522  0.0357  0.1439"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged3 = merged3.sort_values(by=['CHR', 'BP']).reset_index(drop=True)\n",
    "merged3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "5d6b8a07-9f51-4911-a8ac-8fc032199cb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8919817, 8)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "625dd3fc-ce93-4110-93b5-532598923057",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged3.to_csv('../blood_coloc/gwas/RA-10-28-2021/EUR_all_auto-10-2021_processed.txt.gz', compression='gzip', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65af6368-4da9-4ee0-8b26-a5064cbd58c6",
   "metadata": {},
   "source": [
    "## Get index snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "c121a28a-2192-449a-9b60-0433ff9e5e48",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/RA-10-28-2021/EUR_all_auto-10-2021_processed.txt.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "a8a71ee9-8060-42c7-ad15-f4a35b863975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8919817, 8)"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "175dee26-c623-4d46-a6a9-b423c055d8f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>rs202152658</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0588</td>\n",
       "      <td>0.0380</td>\n",
       "      <td>0.1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>751756</td>\n",
       "      <td>rs143225517</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0585</td>\n",
       "      <td>0.0379</td>\n",
       "      <td>0.1230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>rs3094315</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>0.0328</td>\n",
       "      <td>0.1495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>0.2354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>752894</td>\n",
       "      <td>rs3131971</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0522</td>\n",
       "      <td>0.0357</td>\n",
       "      <td>0.1439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR      BP          SNP alt ref    Beta      SE    Pval\n",
       "0    1  751343  rs202152658   A   T -0.0588  0.0380  0.1220\n",
       "1    1  751756  rs143225517   C   T -0.0585  0.0379  0.1230\n",
       "2    1  752566    rs3094315   A   G  0.0472  0.0328  0.1495\n",
       "3    1  752721    rs3131972   G   A  0.0325  0.0274  0.2354\n",
       "4    1  752894    rs3131971   C   T  0.0522  0.0357  0.1439"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "534af1c8-8f86-4f63-a6d3-8517f2173264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29823, 8)\n"
     ]
    }
   ],
   "source": [
    "pval_thres = 5e-8\n",
    "sig = raw.loc[raw['Pval']<pval_thres].reset_index(drop=True)\n",
    "print(sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "9cebbd71-e8e6-43b7-a7c7-9a5d77b2016d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 29823/29823 [00:28<00:00, 1032.76it/s]\n"
     ]
    }
   ],
   "source": [
    "sumstat = sig.copy()\n",
    "sumstat['cyto'] = None\n",
    "for i, row in tqdm.tqdm(sumstat.iterrows(), total=sumstat.shape[0]):\n",
    "    tmp = cytoBand.loc[(cytoBand[0]==f'chr{row.CHR}') & (cytoBand[1]<=row.BP) & (cytoBand[2]>=row.BP)]\n",
    "    sumstat.iloc[i,8] = f'{row.CHR}{tmp[3].values[0]}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "ac197810-1377-4ad4-ae89-fc1337e2825d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(69, 9)\n"
     ]
    }
   ],
   "source": [
    "index_df = sumstat.loc[sumstat.groupby('cyto')['Pval'].idxmin()].sort_values(by=['CHR', 'BP']).reset_index(drop=True)\n",
    "index_df = index_df.loc[index_df['SNP'].str.startswith('rs')]\n",
    "print(index_df.shape)\n",
    "index_df.to_csv('../blood_coloc/gwas/RA-10-28-2021/EUR_all_auto_indexsnp.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "1820ebbd-3c83-4b02-b1f4-5fab69b6090f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pval</th>\n",
       "      <th>cyto</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2483961</td>\n",
       "      <td>rs2258734</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0939</td>\n",
       "      <td>0.0148</td>\n",
       "      <td>2.521000e-10</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2501222</td>\n",
       "      <td>rs10797431</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>7.040000e-10</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2516781</td>\n",
       "      <td>rs60733400</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0880</td>\n",
       "      <td>0.0136</td>\n",
       "      <td>1.127000e-10</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2520302</td>\n",
       "      <td>rs6667564</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0844</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>3.228000e-10</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2520500</td>\n",
       "      <td>rs6701238</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0843</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>3.307000e-10</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       BP         SNP alt ref    Beta      SE          Pval     cyto\n",
       "0    1  2483961   rs2258734   A   G -0.0939  0.0148  2.521000e-10  1p36.32\n",
       "1    1  2501222  rs10797431   T   G -0.0852  0.0138  7.040000e-10  1p36.32\n",
       "2    1  2516781  rs60733400   A   G -0.0880  0.0136  1.127000e-10  1p36.32\n",
       "3    1  2520302   rs6667564   C   T -0.0844  0.0134  3.228000e-10  1p36.32\n",
       "4    1  2520500   rs6701238   A   G -0.0843  0.0134  3.307000e-10  1p36.32"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b84bc26-c264-4401-a388-6868f2233ce1",
   "metadata": {},
   "source": [
    "# MS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ef8dd89-2b0c-4a5d-878e-93edf79575b1",
   "metadata": {},
   "source": [
    "## GWAS preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "f33569fa-80d3-4d4b-989a-dda9a2557eff",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/MS_meta', sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "e3b8de53-e2ab-4042-84f8-ee9c5551ff4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8589719, 8)"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "35f4a782-16cf-4bf2-86e3-4c88bc374dd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>OR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11154</td>\n",
       "      <td>chr1:11154</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>0.7911</td>\n",
       "      <td>0.9818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>11565</td>\n",
       "      <td>chr1:11565</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>0.8735</td>\n",
       "      <td>0.9924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11710</td>\n",
       "      <td>chr1:11710</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "      <td>0.7793</td>\n",
       "      <td>0.9806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>11961</td>\n",
       "      <td>chr1:11961</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "      <td>0.7953</td>\n",
       "      <td>0.9817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>12078</td>\n",
       "      <td>chr1:12078</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR     BP         SNP effect_allele other_allele  N       P      OR\n",
       "0    1  11154  chr1:11154             C            A  4  0.7911  0.9818\n",
       "1    1  11565  chr1:11565             G            T  5  0.8735  0.9924\n",
       "2    1  11710  chr1:11710             T            G  4  0.7793  0.9806\n",
       "3    1  11961  chr1:11961             A            G  4  0.7953  0.9817\n",
       "4    1  12078  chr1:12078             G            A  1     NaN     NaN"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "55393f93-ff87-413d-89ed-bc0839059638",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>MAF</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs201725126</td>\n",
       "      <td>1</td>\n",
       "      <td>13116</td>\n",
       "      <td>0.186879</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs200579949</td>\n",
       "      <td>1</td>\n",
       "      <td>13118</td>\n",
       "      <td>0.186879</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs75454623</td>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>0.479125</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs199856693</td>\n",
       "      <td>1</td>\n",
       "      <td>14933</td>\n",
       "      <td>0.050696</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs78601809</td>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>0.268390</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP  CHR     BP       MAF A1 A2\n",
       "0  rs201725126    1  13116  0.186879  T  G\n",
       "1  rs200579949    1  13118  0.186879  A  G\n",
       "2   rs75454623    1  14930  0.479125  A  G\n",
       "3  rs199856693    1  14933  0.050696  G  A\n",
       "4   rs78601809    1  15211  0.268390  T  G"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "a2ec3cc4-023c-4255-9245-4c7a86fc34c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw.rename(columns={'A1': 'effect_allele', 'A2': 'other_allele'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "263bbf4d-0a9b-485d-8740-4df62e6a2aa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((7959516, 12), (8589719, 8), (9667224, 6))"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(raw, ref, on=['CHR', 'BP'], how='inner')\n",
    "merged.shape, raw.shape, ref.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "fc771953-f2b0-41c9-9d7e-872a042295de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7959422, 12)"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = merged.loc[((merged.effect_allele==merged.A1) & (merged.other_allele==merged.A2)) | ((merged.effect_allele==merged.A2) & (merged.other_allele==merged.A1))]\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "a0e3c239-7eb6-49a8-bfa5-58b407ef9a73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7753386, 12)"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.dropna(subset=['P', 'OR'], inplace=True)\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "da96b0e2-d14d-4e61-a5aa-9f972f24e39a",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged['beta'] = np.log(merged.OR)\n",
    "merged['se'] = np.abs(merged['beta'] / ss.norm.ppf(merged.P/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "c8c94bdd-7e36-4bb8-89d4-9343c1cd6f87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7753386, 10)"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged2 = merged[['CHR', 'BP', 'SNP_y', 'effect_allele', 'other_allele', 'beta', 'se', 'OR', 'P', 'N']].sort_values(by=['CHR', 'BP']).reset_index(drop=True)\n",
    "merged2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "18da1647-ae43-4f0a-8d6f-2bdf0ad9dc2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged2.rename(columns={'SNP_y': 'SNP'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "4ef5d0ea-8eea-412c-a62b-597729e34f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged2.to_csv('../blood_coloc/gwas/MS_meta_processed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b3814a9-a6c5-4d78-939d-8f0e8ef3843d",
   "metadata": {},
   "source": [
    "## Get index snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "c3f78f8c-baad-4883-acaa-f68aa77f42ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7753386, 10)"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/MS_meta_processed', sep='\\t')\n",
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "70638e62-e07d-406c-a0c6-2e3b68643552",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23262, 10)\n"
     ]
    }
   ],
   "source": [
    "pval_thres = 5e-8\n",
    "sig = raw.loc[raw['P']<pval_thres].reset_index(drop=True)\n",
    "print(sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "d48518ce-6dbb-4acd-9936-9ccb14ee48c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2470848</td>\n",
       "      <td>rs12042279</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.099158</td>\n",
       "      <td>0.018130</td>\n",
       "      <td>0.9056</td>\n",
       "      <td>4.518000e-08</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2472081</td>\n",
       "      <td>rs1886731</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.105583</td>\n",
       "      <td>0.018366</td>\n",
       "      <td>0.8998</td>\n",
       "      <td>8.982000e-09</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2472144</td>\n",
       "      <td>rs4487972</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.099378</td>\n",
       "      <td>0.018091</td>\n",
       "      <td>0.9054</td>\n",
       "      <td>3.944000e-08</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2474205</td>\n",
       "      <td>rs942824</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.100815</td>\n",
       "      <td>0.018060</td>\n",
       "      <td>0.9041</td>\n",
       "      <td>2.374000e-08</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2475212</td>\n",
       "      <td>rs10797429</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.101037</td>\n",
       "      <td>0.018050</td>\n",
       "      <td>0.9039</td>\n",
       "      <td>2.172000e-08</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       BP         SNP effect_allele other_allele      beta        se  \\\n",
       "0    1  2470848  rs12042279             A            G -0.099158  0.018130   \n",
       "1    1  2472081   rs1886731             C            T -0.105583  0.018366   \n",
       "2    1  2472144   rs4487972             C            G -0.099378  0.018091   \n",
       "3    1  2474205    rs942824             C            T -0.100815  0.018060   \n",
       "4    1  2475212  rs10797429             C            G -0.101037  0.018050   \n",
       "\n",
       "       OR             P   N  \n",
       "0  0.9056  4.518000e-08  15  \n",
       "1  0.8998  8.982000e-09  15  \n",
       "2  0.9054  3.944000e-08  15  \n",
       "3  0.9041  2.374000e-08  15  \n",
       "4  0.9039  2.172000e-08  15  "
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "de5a04bd-7c05-410e-8f1f-45394f9a6630",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 23262/23262 [00:22<00:00, 1014.44it/s]\n"
     ]
    }
   ],
   "source": [
    "sumstat = sig.copy()\n",
    "sumstat['cyto'] = None\n",
    "for i, row in tqdm.tqdm(sumstat.iterrows(), total=sumstat.shape[0]):\n",
    "    tmp = cytoBand.loc[(cytoBand[0]==f'chr{row.CHR}') & (cytoBand[1]<=row.BP) & (cytoBand[2]>=row.BP)]\n",
    "    sumstat.iloc[i,10] = f'{row.CHR}{tmp[3].values[0]}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "6d70a1c4-d920-4a6e-9a32-69399142ea6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(77, 11)\n"
     ]
    }
   ],
   "source": [
    "index_df = sumstat.loc[sumstat.groupby('cyto')['P'].idxmin()].sort_values(by=['CHR', 'BP']).reset_index(drop=True)\n",
    "index_df = index_df.loc[index_df['SNP'].str.startswith('rs')]\n",
    "print(index_df.shape)\n",
    "index_df.to_csv('../blood_coloc/gwas/MS_meta_indexsnp.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "216dfe06-c766-42b6-b4a1-15db211619c6",
   "metadata": {},
   "source": [
    "# Asthma"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4a2a4b1-1672-461c-8b89-a1bae33e0cd1",
   "metadata": {},
   "source": [
    "## GWAS preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "42ecfa37-3eb7-4cf4-a5e1-f11cd5a9e99f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13791467, 12)"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/20002_1111.gwas.imputed_v3.both_sexes.tsv.bgz', compression='gzip', sep='\\t')\n",
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "e0c3f8c1-0c67-4ba8-8565-f6b49045beb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>expected_case_minor_AC</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.429410e-09</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>224.209000</td>\n",
       "      <td>81.598800</td>\n",
       "      <td>2.747710</td>\n",
       "      <td>0.006002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.749750e-06</td>\n",
       "      <td>0.482220</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>0.011765</td>\n",
       "      <td>-0.115130</td>\n",
       "      <td>0.159834</td>\n",
       "      <td>-0.720309</td>\n",
       "      <td>0.471335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.877650e-04</td>\n",
       "      <td>15.747500</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>135.620000</td>\n",
       "      <td>15.419600</td>\n",
       "      <td>-0.002088</td>\n",
       "      <td>0.028666</td>\n",
       "      <td>-0.072850</td>\n",
       "      <td>0.941926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.662880e-06</td>\n",
       "      <td>0.474934</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.116331</td>\n",
       "      <td>0.159839</td>\n",
       "      <td>-0.727802</td>\n",
       "      <td>0.466735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106370e-01</td>\n",
       "      <td>9278.860000</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>79910.800000</td>\n",
       "      <td>9184.790000</td>\n",
       "      <td>-0.001472</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>-1.111400</td>\n",
       "      <td>0.266398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant minor_allele      minor_AF  expected_case_minor_AC  \\\n",
       "0    1:15791:C:T            T  5.429410e-09                0.000455   \n",
       "1    1:69487:G:A            A  5.749750e-06                0.482220   \n",
       "2    1:69569:T:C            C  1.877650e-04               15.747500   \n",
       "3   1:139853:C:T            T  5.662880e-06                0.474934   \n",
       "4  1:692794:CA:C            C  1.106370e-01             9278.860000   \n",
       "\n",
       "   low_confidence_variant  n_complete_samples            AC          ytx  \\\n",
       "0                    True              361141      0.003922     0.003922   \n",
       "1                    True              361141      4.152940     0.011765   \n",
       "2                    True              361141    135.620000    15.419600   \n",
       "3                    True              361141      4.090200     0.000000   \n",
       "4                   False              361141  79910.800000  9184.790000   \n",
       "\n",
       "         beta         se     tstat      pval  \n",
       "0  224.209000  81.598800  2.747710  0.006002  \n",
       "1   -0.115130   0.159834 -0.720309  0.471335  \n",
       "2   -0.002088   0.028666 -0.072850  0.941926  \n",
       "3   -0.116331   0.159839 -0.727802  0.466735  \n",
       "4   -0.001472   0.001324 -1.111400  0.266398  "
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff6c860-8b13-4bf5-9472-7d5fdf34f344",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw[['CHR', 'BP', 'ref', 'alt']] = raw.SNP.str.split('_', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "b95af450-de8c-4d93-86cd-dca4dc9f0a58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13791467, 17)"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(raw, ukb_variants[['variant', 'chr', 'pos', 'ref', 'alt', 'rsid']], on='variant', how='inner')\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "92a7313b-bdb3-4118-81a9-9e12c661ec2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>expected_case_minor_AC</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.429410e-09</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>224.209000</td>\n",
       "      <td>81.598800</td>\n",
       "      <td>2.747710</td>\n",
       "      <td>0.006002</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs547522712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.749750e-06</td>\n",
       "      <td>0.482220</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>0.011765</td>\n",
       "      <td>-0.115130</td>\n",
       "      <td>0.159834</td>\n",
       "      <td>-0.720309</td>\n",
       "      <td>0.471335</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs568226429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.877650e-04</td>\n",
       "      <td>15.747500</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>135.620000</td>\n",
       "      <td>15.419600</td>\n",
       "      <td>-0.002088</td>\n",
       "      <td>0.028666</td>\n",
       "      <td>-0.072850</td>\n",
       "      <td>0.941926</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs2531267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.662880e-06</td>\n",
       "      <td>0.474934</td>\n",
       "      <td>True</td>\n",
       "      <td>361141</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.116331</td>\n",
       "      <td>0.159839</td>\n",
       "      <td>-0.727802</td>\n",
       "      <td>0.466735</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs533633326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106370e-01</td>\n",
       "      <td>9278.860000</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>79910.800000</td>\n",
       "      <td>9184.790000</td>\n",
       "      <td>-0.001472</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>-1.111400</td>\n",
       "      <td>0.266398</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>1:692794_CA_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant minor_allele      minor_AF  expected_case_minor_AC  \\\n",
       "0    1:15791:C:T            T  5.429410e-09                0.000455   \n",
       "1    1:69487:G:A            A  5.749750e-06                0.482220   \n",
       "2    1:69569:T:C            C  1.877650e-04               15.747500   \n",
       "3   1:139853:C:T            T  5.662880e-06                0.474934   \n",
       "4  1:692794:CA:C            C  1.106370e-01             9278.860000   \n",
       "\n",
       "   low_confidence_variant  n_complete_samples            AC          ytx  \\\n",
       "0                    True              361141      0.003922     0.003922   \n",
       "1                    True              361141      4.152940     0.011765   \n",
       "2                    True              361141    135.620000    15.419600   \n",
       "3                    True              361141      4.090200     0.000000   \n",
       "4                   False              361141  79910.800000  9184.790000   \n",
       "\n",
       "         beta         se     tstat      pval chr     pos ref alt  \\\n",
       "0  224.209000  81.598800  2.747710  0.006002   1   15791   C   T   \n",
       "1   -0.115130   0.159834 -0.720309  0.471335   1   69487   G   A   \n",
       "2   -0.002088   0.028666 -0.072850  0.941926   1   69569   T   C   \n",
       "3   -0.116331   0.159839 -0.727802  0.466735   1  139853   C   T   \n",
       "4   -0.001472   0.001324 -1.111400  0.266398   1  692794  CA   C   \n",
       "\n",
       "            rsid  \n",
       "0    rs547522712  \n",
       "1    rs568226429  \n",
       "2      rs2531267  \n",
       "3    rs533633326  \n",
       "4  1:692794_CA_C  "
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "2a27bc69-06d8-4016-a54e-0c5d524a4bb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13336576, 17)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(13135953, 17)"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = merged.loc[merged.chr.isin(np.arange(1,23))].astype({'chr': int})\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged.low_confidence_variant==False].sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "a8145925-b414-4636-993d-2669b0363fff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>expected_case_minor_AC</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.110637</td>\n",
       "      <td>9278.860</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>79910.80</td>\n",
       "      <td>9184.790</td>\n",
       "      <td>-0.001472</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>-1.111400</td>\n",
       "      <td>0.266398</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>1:692794_CA_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.115823</td>\n",
       "      <td>9713.840</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>83656.80</td>\n",
       "      <td>9665.270</td>\n",
       "      <td>-0.000659</td>\n",
       "      <td>0.001251</td>\n",
       "      <td>-0.526387</td>\n",
       "      <td>0.598620</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs12238997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.097298</td>\n",
       "      <td>8160.220</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>70276.90</td>\n",
       "      <td>8086.070</td>\n",
       "      <td>-0.001301</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>-0.925050</td>\n",
       "      <td>0.354940</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs371890604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.015686</td>\n",
       "      <td>1315.590</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>11330.00</td>\n",
       "      <td>1310.450</td>\n",
       "      <td>-0.000595</td>\n",
       "      <td>0.003356</td>\n",
       "      <td>-0.177415</td>\n",
       "      <td>0.859182</td>\n",
       "      <td>1</td>\n",
       "      <td>717587</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs144155419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>145.379</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>1252.02</td>\n",
       "      <td>157.067</td>\n",
       "      <td>0.011290</td>\n",
       "      <td>0.009901</td>\n",
       "      <td>1.140270</td>\n",
       "      <td>0.254175</td>\n",
       "      <td>1</td>\n",
       "      <td>723329</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>rs189787166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant minor_allele  minor_AF  expected_case_minor_AC  \\\n",
       "0  1:692794:CA:C            C  0.110637                9278.860   \n",
       "1   1:693731:A:G            G  0.115823                9713.840   \n",
       "2   1:707522:G:C            C  0.097298                8160.220   \n",
       "3   1:717587:G:A            A  0.015686                1315.590   \n",
       "4   1:723329:A:T            T  0.001733                 145.379   \n",
       "\n",
       "   low_confidence_variant  n_complete_samples        AC       ytx      beta  \\\n",
       "0                   False              361141  79910.80  9184.790 -0.001472   \n",
       "1                   False              361141  83656.80  9665.270 -0.000659   \n",
       "2                   False              361141  70276.90  8086.070 -0.001301   \n",
       "3                   False              361141  11330.00  1310.450 -0.000595   \n",
       "4                   False              361141   1252.02   157.067  0.011290   \n",
       "\n",
       "         se     tstat      pval  chr     pos ref alt           rsid  \n",
       "0  0.001324 -1.111400  0.266398    1  692794  CA   C  1:692794_CA_C  \n",
       "1  0.001251 -0.526387  0.598620    1  693731   A   G     rs12238997  \n",
       "2  0.001406 -0.925050  0.354940    1  707522   G   C    rs371890604  \n",
       "3  0.003356 -0.177415  0.859182    1  717587   G   A    rs144155419  \n",
       "4  0.009901  1.140270  0.254175    1  723329   A   T    rs189787166  "
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "2237e182-2854-4436-a089-1d47d425e7a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12556648, 17)"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = merged.loc[merged['rsid'].str.startswith('rs')]\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "c7a9345f-4643-4ec0-a0bf-580d63374878",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>expected_case_minor_AC</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.115823</td>\n",
       "      <td>9713.840</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>83656.80</td>\n",
       "      <td>9665.270</td>\n",
       "      <td>-0.000659</td>\n",
       "      <td>0.001251</td>\n",
       "      <td>-0.526387</td>\n",
       "      <td>0.598620</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs12238997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.097298</td>\n",
       "      <td>8160.220</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>70276.90</td>\n",
       "      <td>8086.070</td>\n",
       "      <td>-0.001301</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>-0.925050</td>\n",
       "      <td>0.354940</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs371890604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.015686</td>\n",
       "      <td>1315.590</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>11330.00</td>\n",
       "      <td>1310.450</td>\n",
       "      <td>-0.000595</td>\n",
       "      <td>0.003356</td>\n",
       "      <td>-0.177415</td>\n",
       "      <td>0.859182</td>\n",
       "      <td>1</td>\n",
       "      <td>717587</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs144155419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>145.379</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>1252.02</td>\n",
       "      <td>157.067</td>\n",
       "      <td>0.011290</td>\n",
       "      <td>0.009901</td>\n",
       "      <td>1.140270</td>\n",
       "      <td>0.254175</td>\n",
       "      <td>1</td>\n",
       "      <td>723329</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>rs189787166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:730087:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.056457</td>\n",
       "      <td>4734.970</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>40778.20</td>\n",
       "      <td>4673.710</td>\n",
       "      <td>-0.001722</td>\n",
       "      <td>0.001743</td>\n",
       "      <td>-0.988276</td>\n",
       "      <td>0.323018</td>\n",
       "      <td>1</td>\n",
       "      <td>730087</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs148120343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant minor_allele  minor_AF  expected_case_minor_AC  \\\n",
       "1  1:693731:A:G            G  0.115823                9713.840   \n",
       "2  1:707522:G:C            C  0.097298                8160.220   \n",
       "3  1:717587:G:A            A  0.015686                1315.590   \n",
       "4  1:723329:A:T            T  0.001733                 145.379   \n",
       "5  1:730087:T:C            C  0.056457                4734.970   \n",
       "\n",
       "   low_confidence_variant  n_complete_samples        AC       ytx      beta  \\\n",
       "1                   False              361141  83656.80  9665.270 -0.000659   \n",
       "2                   False              361141  70276.90  8086.070 -0.001301   \n",
       "3                   False              361141  11330.00  1310.450 -0.000595   \n",
       "4                   False              361141   1252.02   157.067  0.011290   \n",
       "5                   False              361141  40778.20  4673.710 -0.001722   \n",
       "\n",
       "         se     tstat      pval  chr     pos ref alt         rsid  \n",
       "1  0.001251 -0.526387  0.598620    1  693731   A   G   rs12238997  \n",
       "2  0.001406 -0.925050  0.354940    1  707522   G   C  rs371890604  \n",
       "3  0.003356 -0.177415  0.859182    1  717587   G   A  rs144155419  \n",
       "4  0.009901  1.140270  0.254175    1  723329   A   T  rs189787166  \n",
       "5  0.001743 -0.988276  0.323018    1  730087   T   C  rs148120343  "
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "7a41923e-0399-4d67-a060-7a2f7721a5f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[['rsid', 'chr', 'pos', 'ref', 'alt', 'minor_AF', 'beta', 'se', 'tstat', 'pval']].to_csv('../blood_coloc/gwas/Asthma_ukb.csv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9be69064-8176-4cec-8114-bfe89dcebf81",
   "metadata": {},
   "source": [
    "## Get index snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "199dbbab-cafe-4b60-946b-c329f152c7a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12556648, 10)"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/Asthma_ukb.csv', sep='\\t')\n",
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "6ca336bf-3f64-4581-89b9-e1855cf4a85a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs12238997</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.115823</td>\n",
       "      <td>-0.000659</td>\n",
       "      <td>0.001251</td>\n",
       "      <td>-0.526387</td>\n",
       "      <td>0.598620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs371890604</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.097298</td>\n",
       "      <td>-0.001301</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>-0.925050</td>\n",
       "      <td>0.354940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs144155419</td>\n",
       "      <td>1</td>\n",
       "      <td>717587</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.015686</td>\n",
       "      <td>-0.000595</td>\n",
       "      <td>0.003356</td>\n",
       "      <td>-0.177415</td>\n",
       "      <td>0.859182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs189787166</td>\n",
       "      <td>1</td>\n",
       "      <td>723329</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>0.011290</td>\n",
       "      <td>0.009901</td>\n",
       "      <td>1.140270</td>\n",
       "      <td>0.254175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs148120343</td>\n",
       "      <td>1</td>\n",
       "      <td>730087</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.056457</td>\n",
       "      <td>-0.001722</td>\n",
       "      <td>0.001743</td>\n",
       "      <td>-0.988276</td>\n",
       "      <td>0.323018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          rsid  chr     pos ref alt  minor_AF      beta        se     tstat  \\\n",
       "0   rs12238997    1  693731   A   G  0.115823 -0.000659  0.001251 -0.526387   \n",
       "1  rs371890604    1  707522   G   C  0.097298 -0.001301  0.001406 -0.925050   \n",
       "2  rs144155419    1  717587   G   A  0.015686 -0.000595  0.003356 -0.177415   \n",
       "3  rs189787166    1  723329   A   T  0.001733  0.011290  0.009901  1.140270   \n",
       "4  rs148120343    1  730087   T   C  0.056457 -0.001722  0.001743 -0.988276   \n",
       "\n",
       "       pval  \n",
       "0  0.598620  \n",
       "1  0.354940  \n",
       "2  0.859182  \n",
       "3  0.254175  \n",
       "4  0.323018  "
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "745c9e13-df25-4350-8dcc-0cf21588ac2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22111, 10)\n"
     ]
    }
   ],
   "source": [
    "pval_thres = 5e-8\n",
    "sig = raw.loc[raw['pval']<pval_thres].reset_index(drop=True)\n",
    "print(sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "d06d8eb8-c7d5-4b88-866d-89d520e5aaa0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs159960</td>\n",
       "      <td>1</td>\n",
       "      <td>8476428</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.439602</td>\n",
       "      <td>0.004512</td>\n",
       "      <td>0.000763</td>\n",
       "      <td>5.91554</td>\n",
       "      <td>3.310750e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs301805</td>\n",
       "      <td>1</td>\n",
       "      <td>8481016</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.414070</td>\n",
       "      <td>0.004470</td>\n",
       "      <td>0.000765</td>\n",
       "      <td>5.84052</td>\n",
       "      <td>5.208320e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs301806</td>\n",
       "      <td>1</td>\n",
       "      <td>8482078</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.415936</td>\n",
       "      <td>0.004401</td>\n",
       "      <td>0.000764</td>\n",
       "      <td>5.75747</td>\n",
       "      <td>8.545140e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs301807</td>\n",
       "      <td>1</td>\n",
       "      <td>8484823</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.414515</td>\n",
       "      <td>0.004475</td>\n",
       "      <td>0.000765</td>\n",
       "      <td>5.85053</td>\n",
       "      <td>4.904380e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs301799</td>\n",
       "      <td>1</td>\n",
       "      <td>8489302</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.413459</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.000766</td>\n",
       "      <td>5.77520</td>\n",
       "      <td>7.692560e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       rsid  chr      pos ref alt  minor_AF      beta        se    tstat  \\\n",
       "0  rs159960    1  8476428   A   G  0.439602  0.004512  0.000763  5.91554   \n",
       "1  rs301805    1  8481016   T   G  0.414070  0.004470  0.000765  5.84052   \n",
       "2  rs301806    1  8482078   C   T  0.415936  0.004401  0.000764  5.75747   \n",
       "3  rs301807    1  8484823   A   G  0.414515  0.004475  0.000765  5.85053   \n",
       "4  rs301799    1  8489302   C   T  0.413459  0.004421  0.000766  5.77520   \n",
       "\n",
       "           pval  \n",
       "0  3.310750e-09  \n",
       "1  5.208320e-09  \n",
       "2  8.545140e-09  \n",
       "3  4.904380e-09  \n",
       "4  7.692560e-09  "
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "2c57964c-8129-4b2a-aed3-c202f771734f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 22111/22111 [00:21<00:00, 1035.18it/s]\n"
     ]
    }
   ],
   "source": [
    "sumstat = sig.copy()\n",
    "sumstat['cyto'] = None\n",
    "for i, row in tqdm.tqdm(sumstat.iterrows(), total=sumstat.shape[0]):\n",
    "    tmp = cytoBand.loc[(cytoBand[0]==f'chr{row.chr}') & (cytoBand[1]<=row.pos) & (cytoBand[2]>=row.pos)]\n",
    "    sumstat.iloc[i,10] = f'{row.chr}{tmp[3].values[0]}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "3fba18e0-7a33-4173-8869-3b16b7a190e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(87, 11)\n"
     ]
    }
   ],
   "source": [
    "index_df = sumstat.loc[sumstat.groupby('cyto')['pval'].idxmin()].sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "index_df = index_df.loc[index_df['rsid'].str.startswith('rs')]\n",
    "print(index_df.shape)\n",
    "index_df.to_csv('../blood_coloc/gwas/Asthma_ukb_indexsnp.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e34580a-2612-4f5a-8a9f-e1b00988aca7",
   "metadata": {},
   "source": [
    "# Monocyte count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28e59efa-8b7a-413c-813f-6a42d73ecc27",
   "metadata": {},
   "source": [
    "## GWAS preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "41c46740-a308-453b-a6ab-f9cfacb51159",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13791467, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.604550e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>-0.004798</td>\n",
       "      <td>-355.405000</td>\n",
       "      <td>246.614000</td>\n",
       "      <td>-1.441140</td>\n",
       "      <td>0.149547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.935220e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>-0.934619</td>\n",
       "      <td>-0.286586</td>\n",
       "      <td>0.483060</td>\n",
       "      <td>-0.593272</td>\n",
       "      <td>0.552999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.811110e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>126.725000</td>\n",
       "      <td>-7.005580</td>\n",
       "      <td>-0.050922</td>\n",
       "      <td>0.089648</td>\n",
       "      <td>-0.568024</td>\n",
       "      <td>0.570019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.845540e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>-0.937214</td>\n",
       "      <td>-0.286084</td>\n",
       "      <td>0.483074</td>\n",
       "      <td>-0.592215</td>\n",
       "      <td>0.553707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106480e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>77422.100000</td>\n",
       "      <td>-289.390000</td>\n",
       "      <td>-0.003178</td>\n",
       "      <td>0.004064</td>\n",
       "      <td>-0.781944</td>\n",
       "      <td>0.434248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant minor_allele      minor_AF  low_confidence_variant  \\\n",
       "0    1:15791:C:T            T  5.604550e-09                    True   \n",
       "1    1:69487:G:A            A  5.935220e-06                    True   \n",
       "2    1:69569:T:C            C  1.811110e-04                    True   \n",
       "3   1:139853:C:T            T  5.845540e-06                    True   \n",
       "4  1:692794:CA:C            C  1.106480e-01                   False   \n",
       "\n",
       "   n_complete_samples            AC         ytx        beta          se  \\\n",
       "0              349856      0.003922   -0.004798 -355.405000  246.614000   \n",
       "1              349856      4.152940   -0.934619   -0.286586    0.483060   \n",
       "2              349856    126.725000   -7.005580   -0.050922    0.089648   \n",
       "3              349856      4.090200   -0.937214   -0.286084    0.483074   \n",
       "4              349856  77422.100000 -289.390000   -0.003178    0.004064   \n",
       "\n",
       "      tstat      pval  \n",
       "0 -1.441140  0.149547  \n",
       "1 -0.593272  0.552999  \n",
       "2 -0.568024  0.570019  \n",
       "3 -0.592215  0.553707  \n",
       "4 -0.781944  0.434248  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/30130_irnt.gwas.imputed_v3.both_sexes.tsv.bgz', compression='gzip', sep='\\t')\n",
    "print(raw.shape)\n",
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9fcb372f-0c8a-4599-be6d-c18f5a9847fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "      <th>varid</th>\n",
       "      <th>consequence</th>\n",
       "      <th>consequence_category</th>\n",
       "      <th>info</th>\n",
       "      <th>...</th>\n",
       "      <th>p_hwe</th>\n",
       "      <th>n_called</th>\n",
       "      <th>n_not_called</th>\n",
       "      <th>n_hom_ref</th>\n",
       "      <th>n_het</th>\n",
       "      <th>n_hom_var</th>\n",
       "      <th>n_non_ref</th>\n",
       "      <th>r_heterozygosity</th>\n",
       "      <th>r_het_hom_var</th>\n",
       "      <th>r_expected_het_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs547522712</td>\n",
       "      <td>1:15791_C_T</td>\n",
       "      <td>splice_region_variant</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.861678</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>361194</td>\n",
       "      <td>0</td>\n",
       "      <td>361194</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs568226429</td>\n",
       "      <td>1:69487_G_A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.956975</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500004</td>\n",
       "      <td>361194</td>\n",
       "      <td>0</td>\n",
       "      <td>361190</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs2531267</td>\n",
       "      <td>1:69569_T_C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.831664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.506315</td>\n",
       "      <td>361194</td>\n",
       "      <td>0</td>\n",
       "      <td>361058</td>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>136</td>\n",
       "      <td>0.000377</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs533633326</td>\n",
       "      <td>1:139853_C_T</td>\n",
       "      <td>splice_region_variant</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.985255</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500004</td>\n",
       "      <td>361194</td>\n",
       "      <td>0</td>\n",
       "      <td>361190</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>1:692794_CA_C</td>\n",
       "      <td>1:692794_CA_C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>non_coding</td>\n",
       "      <td>0.824483</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031437</td>\n",
       "      <td>361188</td>\n",
       "      <td>6</td>\n",
       "      <td>286823</td>\n",
       "      <td>70224</td>\n",
       "      <td>4141</td>\n",
       "      <td>74365</td>\n",
       "      <td>0.194425</td>\n",
       "      <td>16.9582</td>\n",
       "      <td>0.193734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant chr     pos ref alt           rsid          varid  \\\n",
       "0    1:15791:C:T   1   15791   C   T    rs547522712    1:15791_C_T   \n",
       "1    1:69487:G:A   1   69487   G   A    rs568226429    1:69487_G_A   \n",
       "2    1:69569:T:C   1   69569   T   C      rs2531267    1:69569_T_C   \n",
       "3   1:139853:C:T   1  139853   C   T    rs533633326   1:139853_C_T   \n",
       "4  1:692794:CA:C   1  692794  CA   C  1:692794_CA_C  1:692794_CA_C   \n",
       "\n",
       "             consequence consequence_category      info  ...     p_hwe  \\\n",
       "0  splice_region_variant             missense  0.861678  ...  0.500000   \n",
       "1       missense_variant             missense  0.956975  ...  0.500004   \n",
       "2       missense_variant             missense  0.831664  ...  0.506315   \n",
       "3  splice_region_variant             missense  0.985255  ...  0.500004   \n",
       "4  upstream_gene_variant           non_coding  0.824483  ...  0.031437   \n",
       "\n",
       "   n_called  n_not_called n_hom_ref  n_het  n_hom_var  n_non_ref  \\\n",
       "0    361194             0    361194      0          0          0   \n",
       "1    361194             0    361190      4          0          4   \n",
       "2    361194             0    361058    136          0        136   \n",
       "3    361194             0    361190      4          0          4   \n",
       "4    361188             6    286823  70224       4141      74365   \n",
       "\n",
       "   r_heterozygosity  r_het_hom_var  r_expected_het_frequency  \n",
       "0          0.000000            NaN                  0.000000  \n",
       "1          0.000011            NaN                  0.000011  \n",
       "2          0.000377            NaN                  0.000376  \n",
       "3          0.000011            NaN                  0.000011  \n",
       "4          0.194425        16.9582                  0.193734  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ukb_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bdcc23f9-8edb-470d-95ab-6350e1dbc5f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13791467, 16)\n",
      "(13336576, 16)\n",
      "(13135652, 16)\n",
      "(12556367, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.115822</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>81042.30</td>\n",
       "      <td>-22.89850</td>\n",
       "      <td>0.001750</td>\n",
       "      <td>0.003841</td>\n",
       "      <td>0.455584</td>\n",
       "      <td>0.648689</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs12238997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.097250</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>68047.20</td>\n",
       "      <td>-8.87940</td>\n",
       "      <td>0.002788</td>\n",
       "      <td>0.004319</td>\n",
       "      <td>0.645537</td>\n",
       "      <td>0.518579</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs371890604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.015718</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>10998.10</td>\n",
       "      <td>-3.71433</td>\n",
       "      <td>-0.000652</td>\n",
       "      <td>0.010292</td>\n",
       "      <td>-0.063309</td>\n",
       "      <td>0.949521</td>\n",
       "      <td>1</td>\n",
       "      <td>717587</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs144155419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001720</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>1203.55</td>\n",
       "      <td>-59.28760</td>\n",
       "      <td>-0.058307</td>\n",
       "      <td>0.030551</td>\n",
       "      <td>-1.908550</td>\n",
       "      <td>0.056321</td>\n",
       "      <td>1</td>\n",
       "      <td>723329</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>rs189787166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:730087:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.056479</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>39519.00</td>\n",
       "      <td>-193.37800</td>\n",
       "      <td>-0.004384</td>\n",
       "      <td>0.005348</td>\n",
       "      <td>-0.819718</td>\n",
       "      <td>0.412377</td>\n",
       "      <td>1</td>\n",
       "      <td>730087</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs148120343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant minor_allele  minor_AF  low_confidence_variant  \\\n",
       "1  1:693731:A:G            G  0.115822                   False   \n",
       "2  1:707522:G:C            C  0.097250                   False   \n",
       "3  1:717587:G:A            A  0.015718                   False   \n",
       "4  1:723329:A:T            T  0.001720                   False   \n",
       "5  1:730087:T:C            C  0.056479                   False   \n",
       "\n",
       "   n_complete_samples        AC        ytx      beta        se     tstat  \\\n",
       "1              349856  81042.30  -22.89850  0.001750  0.003841  0.455584   \n",
       "2              349856  68047.20   -8.87940  0.002788  0.004319  0.645537   \n",
       "3              349856  10998.10   -3.71433 -0.000652  0.010292 -0.063309   \n",
       "4              349856   1203.55  -59.28760 -0.058307  0.030551 -1.908550   \n",
       "5              349856  39519.00 -193.37800 -0.004384  0.005348 -0.819718   \n",
       "\n",
       "       pval  chr     pos ref alt         rsid  \n",
       "1  0.648689    1  693731   A   G   rs12238997  \n",
       "2  0.518579    1  707522   G   C  rs371890604  \n",
       "3  0.949521    1  717587   G   A  rs144155419  \n",
       "4  0.056321    1  723329   A   T  rs189787166  \n",
       "5  0.412377    1  730087   T   C  rs148120343  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(raw, ukb_variants[['variant', 'chr', 'pos', 'ref', 'alt', 'rsid']], on='variant', how='inner')\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged.chr.isin(np.arange(1,23))].astype({'chr': int})\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged.low_confidence_variant==False].sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged['rsid'].str.startswith('rs')]\n",
    "print(merged.shape)\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "50653efe-1f6e-427a-8475-aaedd202388b",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[['rsid', 'chr', 'pos', 'ref', 'alt', 'minor_AF', 'beta', 'se', 'tstat', 'pval']].to_csv('../blood_coloc/gwas/Monocyte_count_ukb.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44e7aa6b-16d7-448e-b027-438ba1de8aed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a142b98f-f9c0-497a-a024-b41dc9f7556f",
   "metadata": {},
   "source": [
    "## Get index snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3286976c-30c8-41c9-b831-471b60c0f9c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12556367, 10)\n"
     ]
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/Monocyte_count_ukb.csv', sep='\\t')\n",
    "print(raw.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "35a92205-ccbb-46c3-885f-fbdc2b14c6a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(73361, 10)\n"
     ]
    }
   ],
   "source": [
    "pval_thres = 5e-8\n",
    "sig = raw.loc[raw['pval']<pval_thres].reset_index(drop=True)\n",
    "print(sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3961752e-90a9-43b0-88e1-6459a654f562",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 73361/73361 [01:15<00:00, 972.48it/s]\n"
     ]
    }
   ],
   "source": [
    "sumstat = sig.copy()\n",
    "sumstat['cyto'] = None\n",
    "for i, row in tqdm.tqdm(sumstat.iterrows(), total=sumstat.shape[0]):\n",
    "    tmp = cytoBand.loc[(cytoBand[0]==f'chr{row.chr}') & (cytoBand[1]<=row.pos) & (cytoBand[2]>=row.pos)]\n",
    "    sumstat.iloc[i,10] = f'{row.chr}{tmp[3].values[0]}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "32cfa384-7923-4050-a5cd-c1d0d620f085",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(73361, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>cyto</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs75400926</td>\n",
       "      <td>1</td>\n",
       "      <td>2954378</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.149820</td>\n",
       "      <td>0.018653</td>\n",
       "      <td>0.003376</td>\n",
       "      <td>5.52532</td>\n",
       "      <td>3.291130e-08</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs201056490</td>\n",
       "      <td>1</td>\n",
       "      <td>2954379</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.149820</td>\n",
       "      <td>0.018653</td>\n",
       "      <td>0.003376</td>\n",
       "      <td>5.52532</td>\n",
       "      <td>3.291130e-08</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs147542464</td>\n",
       "      <td>1</td>\n",
       "      <td>2968454</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCAT</td>\n",
       "      <td>0.147834</td>\n",
       "      <td>0.019039</td>\n",
       "      <td>0.003457</td>\n",
       "      <td>5.50733</td>\n",
       "      <td>3.645750e-08</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs717795</td>\n",
       "      <td>1</td>\n",
       "      <td>2970464</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.191292</td>\n",
       "      <td>0.016232</td>\n",
       "      <td>0.002955</td>\n",
       "      <td>5.49220</td>\n",
       "      <td>3.972350e-08</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs2981857</td>\n",
       "      <td>1</td>\n",
       "      <td>2971521</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.191477</td>\n",
       "      <td>0.016183</td>\n",
       "      <td>0.002948</td>\n",
       "      <td>5.48950</td>\n",
       "      <td>4.033510e-08</td>\n",
       "      <td>1p36.32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          rsid  chr      pos ref    alt  minor_AF      beta        se  \\\n",
       "0   rs75400926    1  2954378   T      C  0.149820  0.018653  0.003376   \n",
       "1  rs201056490    1  2954379   T      C  0.149820  0.018653  0.003376   \n",
       "2  rs147542464    1  2968454   C  CCCAT  0.147834  0.019039  0.003457   \n",
       "3     rs717795    1  2970464   T      C  0.191292  0.016232  0.002955   \n",
       "4    rs2981857    1  2971521   T      C  0.191477  0.016183  0.002948   \n",
       "\n",
       "     tstat          pval     cyto  \n",
       "0  5.52532  3.291130e-08  1p36.32  \n",
       "1  5.52532  3.291130e-08  1p36.32  \n",
       "2  5.50733  3.645750e-08  1p36.32  \n",
       "3  5.49220  3.972350e-08  1p36.32  \n",
       "4  5.48950  4.033510e-08  1p36.32  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(sumstat.shape)\n",
    "sumstat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e00199b7-77c4-4675-8b88-9a87aa696e08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(315, 11)\n"
     ]
    }
   ],
   "source": [
    "index_df = sumstat.loc[sumstat.groupby('cyto')['pval'].idxmin()].sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "index_df = index_df.loc[index_df['rsid'].str.startswith('rs')]\n",
    "print(index_df.shape)\n",
    "index_df.to_csv('../blood_coloc/gwas/Monocyte_count_ukb_indexsnp.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "453a3e8b-2b55-4896-a823-d220ca4f20b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bb441e23-3531-4eab-9064-49a22dea4d6b",
   "metadata": {},
   "source": [
    "# Neutrophil count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1f2a9e-44da-487e-a47e-20fc672862a4",
   "metadata": {},
   "source": [
    "## GWAS preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bcafc6c5-498b-4475-89b7-f78f6af34af6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13791467, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.604550e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>-0.002612</td>\n",
       "      <td>-161.915000</td>\n",
       "      <td>253.286000</td>\n",
       "      <td>-0.639257</td>\n",
       "      <td>0.522656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.935220e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>0.419874</td>\n",
       "      <td>0.090707</td>\n",
       "      <td>0.496127</td>\n",
       "      <td>0.182831</td>\n",
       "      <td>0.854931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.811110e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>126.725000</td>\n",
       "      <td>-9.552040</td>\n",
       "      <td>-0.067077</td>\n",
       "      <td>0.092073</td>\n",
       "      <td>-0.728518</td>\n",
       "      <td>0.466297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.845540e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>349856</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>0.402127</td>\n",
       "      <td>0.086197</td>\n",
       "      <td>0.496142</td>\n",
       "      <td>0.173735</td>\n",
       "      <td>0.862073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106480e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>77422.100000</td>\n",
       "      <td>-313.180000</td>\n",
       "      <td>-0.004435</td>\n",
       "      <td>0.004174</td>\n",
       "      <td>-1.062470</td>\n",
       "      <td>0.288024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant minor_allele      minor_AF  low_confidence_variant  \\\n",
       "0    1:15791:C:T            T  5.604550e-09                    True   \n",
       "1    1:69487:G:A            A  5.935220e-06                    True   \n",
       "2    1:69569:T:C            C  1.811110e-04                    True   \n",
       "3   1:139853:C:T            T  5.845540e-06                    True   \n",
       "4  1:692794:CA:C            C  1.106480e-01                   False   \n",
       "\n",
       "   n_complete_samples            AC         ytx        beta          se  \\\n",
       "0              349856      0.003922   -0.002612 -161.915000  253.286000   \n",
       "1              349856      4.152940    0.419874    0.090707    0.496127   \n",
       "2              349856    126.725000   -9.552040   -0.067077    0.092073   \n",
       "3              349856      4.090200    0.402127    0.086197    0.496142   \n",
       "4              349856  77422.100000 -313.180000   -0.004435    0.004174   \n",
       "\n",
       "      tstat      pval  \n",
       "0 -0.639257  0.522656  \n",
       "1  0.182831  0.854931  \n",
       "2 -0.728518  0.466297  \n",
       "3  0.173735  0.862073  \n",
       "4 -1.062470  0.288024  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/30140_irnt.gwas.imputed_v3.both_sexes.tsv.bgz', compression='gzip', sep='\\t')\n",
    "print(raw.shape)\n",
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5c2a22ae-77d9-4ef3-beac-42157eba1055",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13791467, 16)\n",
      "(13336576, 16)\n",
      "(13135652, 16)\n",
      "(12556367, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.115822</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>81042.30</td>\n",
       "      <td>-240.8180</td>\n",
       "      <td>-0.002518</td>\n",
       "      <td>0.003944</td>\n",
       "      <td>-0.638365</td>\n",
       "      <td>0.523236</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs12238997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.097250</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>68047.20</td>\n",
       "      <td>-226.9890</td>\n",
       "      <td>-0.002722</td>\n",
       "      <td>0.004436</td>\n",
       "      <td>-0.613615</td>\n",
       "      <td>0.539470</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs371890604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.015718</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>10998.10</td>\n",
       "      <td>-58.8910</td>\n",
       "      <td>-0.007782</td>\n",
       "      <td>0.010570</td>\n",
       "      <td>-0.736193</td>\n",
       "      <td>0.461614</td>\n",
       "      <td>1</td>\n",
       "      <td>717587</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs144155419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001720</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>1203.55</td>\n",
       "      <td>16.0111</td>\n",
       "      <td>0.018672</td>\n",
       "      <td>0.031377</td>\n",
       "      <td>0.595068</td>\n",
       "      <td>0.551798</td>\n",
       "      <td>1</td>\n",
       "      <td>723329</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>rs189787166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:730087:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.056479</td>\n",
       "      <td>False</td>\n",
       "      <td>349856</td>\n",
       "      <td>39519.00</td>\n",
       "      <td>-28.0723</td>\n",
       "      <td>0.000388</td>\n",
       "      <td>0.005493</td>\n",
       "      <td>0.070610</td>\n",
       "      <td>0.943708</td>\n",
       "      <td>1</td>\n",
       "      <td>730087</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs148120343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant minor_allele  minor_AF  low_confidence_variant  \\\n",
       "1  1:693731:A:G            G  0.115822                   False   \n",
       "2  1:707522:G:C            C  0.097250                   False   \n",
       "3  1:717587:G:A            A  0.015718                   False   \n",
       "4  1:723329:A:T            T  0.001720                   False   \n",
       "5  1:730087:T:C            C  0.056479                   False   \n",
       "\n",
       "   n_complete_samples        AC       ytx      beta        se     tstat  \\\n",
       "1              349856  81042.30 -240.8180 -0.002518  0.003944 -0.638365   \n",
       "2              349856  68047.20 -226.9890 -0.002722  0.004436 -0.613615   \n",
       "3              349856  10998.10  -58.8910 -0.007782  0.010570 -0.736193   \n",
       "4              349856   1203.55   16.0111  0.018672  0.031377  0.595068   \n",
       "5              349856  39519.00  -28.0723  0.000388  0.005493  0.070610   \n",
       "\n",
       "       pval  chr     pos ref alt         rsid  \n",
       "1  0.523236    1  693731   A   G   rs12238997  \n",
       "2  0.539470    1  707522   G   C  rs371890604  \n",
       "3  0.461614    1  717587   G   A  rs144155419  \n",
       "4  0.551798    1  723329   A   T  rs189787166  \n",
       "5  0.943708    1  730087   T   C  rs148120343  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(raw, ukb_variants[['variant', 'chr', 'pos', 'ref', 'alt', 'rsid']], on='variant', how='inner')\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged.chr.isin(np.arange(1,23))].astype({'chr': int})\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged.low_confidence_variant==False].sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "print(merged.shape)\n",
    "merged = merged.loc[merged['rsid'].str.startswith('rs')]\n",
    "print(merged.shape)\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f70bbea4-3819-4186-a085-e89859434fe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[['rsid', 'chr', 'pos', 'ref', 'alt', 'minor_AF', 'beta', 'se', 'tstat', 'pval']].to_csv('../blood_coloc/gwas/Neutrophil_count_ukb.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b97e5af-b256-45f6-969b-3411fd1ec34f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d09f4494-b342-42fc-81d8-c20e400c6a8f",
   "metadata": {},
   "source": [
    "## Get index snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "844759db-ad52-4d8f-831d-98ebe71afc77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12556367, 10)\n"
     ]
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/Neutrophil_count_ukb.csv', sep='\\t')\n",
    "print(raw.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "032dcc9b-f6d3-450b-92ce-1afd59a5ae18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65847, 10)\n"
     ]
    }
   ],
   "source": [
    "pval_thres = 5e-8\n",
    "sig = raw.loc[raw['pval']<pval_thres].reset_index(drop=True)\n",
    "print(sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4d0ac398-9aab-41ac-8429-f36043e26c69",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 65847/65847 [01:39<00:00, 664.14it/s]\n"
     ]
    }
   ],
   "source": [
    "sumstat = sig.copy()\n",
    "sumstat['cyto'] = None\n",
    "for i, row in tqdm.tqdm(sumstat.iterrows(), total=sumstat.shape[0]):\n",
    "    tmp = cytoBand.loc[(cytoBand[0]==f'chr{row.chr}') & (cytoBand[1]<=row.pos) & (cytoBand[2]>=row.pos)]\n",
    "    sumstat.iloc[i,10] = f'{row.chr}{tmp[3].values[0]}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2b6c4402-f582-44c9-86b8-75b848942097",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65847, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>cyto</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs10864351</td>\n",
       "      <td>1</td>\n",
       "      <td>8409672</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.499331</td>\n",
       "      <td>0.014042</td>\n",
       "      <td>0.002390</td>\n",
       "      <td>5.87647</td>\n",
       "      <td>4.194870e-09</td>\n",
       "      <td>1p36.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs8627</td>\n",
       "      <td>1</td>\n",
       "      <td>8412935</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.470298</td>\n",
       "      <td>0.014894</td>\n",
       "      <td>0.002387</td>\n",
       "      <td>6.23845</td>\n",
       "      <td>4.424440e-10</td>\n",
       "      <td>1p36.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1889854</td>\n",
       "      <td>1</td>\n",
       "      <td>8415235</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.220550</td>\n",
       "      <td>0.017697</td>\n",
       "      <td>0.002883</td>\n",
       "      <td>6.13902</td>\n",
       "      <td>8.311910e-10</td>\n",
       "      <td>1p36.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1889853</td>\n",
       "      <td>1</td>\n",
       "      <td>8415290</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.221524</td>\n",
       "      <td>0.017332</td>\n",
       "      <td>0.002877</td>\n",
       "      <td>6.02460</td>\n",
       "      <td>1.696930e-09</td>\n",
       "      <td>1p36.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs2784735</td>\n",
       "      <td>1</td>\n",
       "      <td>8421092</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.459481</td>\n",
       "      <td>0.014153</td>\n",
       "      <td>0.002403</td>\n",
       "      <td>5.89045</td>\n",
       "      <td>3.854950e-09</td>\n",
       "      <td>1p36.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         rsid  chr      pos ref alt  minor_AF      beta        se    tstat  \\\n",
       "0  rs10864351    1  8409672   C   A  0.499331  0.014042  0.002390  5.87647   \n",
       "1      rs8627    1  8412935   C   T  0.470298  0.014894  0.002387  6.23845   \n",
       "2   rs1889854    1  8415235   C   T  0.220550  0.017697  0.002883  6.13902   \n",
       "3   rs1889853    1  8415290   C   A  0.221524  0.017332  0.002877  6.02460   \n",
       "4   rs2784735    1  8421092   C   T  0.459481  0.014153  0.002403  5.89045   \n",
       "\n",
       "           pval     cyto  \n",
       "0  4.194870e-09  1p36.23  \n",
       "1  4.424440e-10  1p36.23  \n",
       "2  8.311910e-10  1p36.23  \n",
       "3  1.696930e-09  1p36.23  \n",
       "4  3.854950e-09  1p36.23  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(sumstat.shape)\n",
    "sumstat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f5869bb6-0af5-432a-a5be-735c9342bc3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(285, 11)\n"
     ]
    }
   ],
   "source": [
    "index_df = sumstat.loc[sumstat.groupby('cyto')['pval'].idxmin()].sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "index_df = index_df.loc[index_df['rsid'].str.startswith('rs')]\n",
    "print(index_df.shape)\n",
    "index_df.to_csv('../blood_coloc/gwas/Neutrophil_count_ukb_indexsnp.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "269cd4e2-c4ad-4e54-8c16-8e7a13902e1c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e2e0275a-156a-421c-9a9e-83d81474ad65",
   "metadata": {},
   "source": [
    "# IBD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c32636df-73b1-463c-8deb-d3d1e2c17a7a",
   "metadata": {},
   "source": [
    "## GWAS preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "c05a3f0b-8135-4c5e-a857-e5f75ddf8c28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9735446, 15)"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/ibd_build37_59957_20161107.txt.gz', sep='\\t', compression='gzip')\n",
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f20eb962-85f8-490a-ab70-da2c9c2c490d",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw[['CHR-BP', 'ref', 'alt']] = raw.MarkerName.str.split('_', expand=True)\n",
    "raw[['CHR', 'BP']] = raw['CHR-BP'].str.split(':', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "9262688f-032e-4dec-805a-2ef2f7cc3666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P.value</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>HetPVal</th>\n",
       "      <th>Pval_IBDseq</th>\n",
       "      <th>Pval_IIBDGC</th>\n",
       "      <th>Pval_GWAS3</th>\n",
       "      <th>Min_single_cohort_pval</th>\n",
       "      <th>CHR-BP</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:100000012_G_T</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0078</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.5950</td>\n",
       "      <td>+-+</td>\n",
       "      <td>82.7</td>\n",
       "      <td>11.592</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00304</td>\n",
       "      <td>0.239590</td>\n",
       "      <td>0.2746</td>\n",
       "      <td>0.002291</td>\n",
       "      <td>0.002291</td>\n",
       "      <td>1:100000012</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>100000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:10000006_G_A</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.1109</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>+--</td>\n",
       "      <td>54.3</td>\n",
       "      <td>4.373</td>\n",
       "      <td>2</td>\n",
       "      <td>0.11230</td>\n",
       "      <td>0.054628</td>\n",
       "      <td>0.9403</td>\n",
       "      <td>0.404895</td>\n",
       "      <td>0.054628</td>\n",
       "      <td>1:10000006</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>10000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:100000827_C_T</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0144</td>\n",
       "      <td>0.0136</td>\n",
       "      <td>0.2915</td>\n",
       "      <td>+-+</td>\n",
       "      <td>72.4</td>\n",
       "      <td>7.253</td>\n",
       "      <td>2</td>\n",
       "      <td>0.02661</td>\n",
       "      <td>0.499568</td>\n",
       "      <td>0.6470</td>\n",
       "      <td>0.005517</td>\n",
       "      <td>0.005517</td>\n",
       "      <td>1:100000827</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>100000827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:100000843_T_C</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>0.0289</td>\n",
       "      <td>0.1954</td>\n",
       "      <td>+++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.575</td>\n",
       "      <td>2</td>\n",
       "      <td>0.75030</td>\n",
       "      <td>0.387836</td>\n",
       "      <td>0.2238</td>\n",
       "      <td>0.869664</td>\n",
       "      <td>0.223800</td>\n",
       "      <td>1:100000843</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>100000843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:100000989_A_ATC</td>\n",
       "      <td>a</td>\n",
       "      <td>atc</td>\n",
       "      <td>0.0431</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>0.1499</td>\n",
       "      <td>+++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.835</td>\n",
       "      <td>2</td>\n",
       "      <td>0.65860</td>\n",
       "      <td>0.290844</td>\n",
       "      <td>0.1833</td>\n",
       "      <td>0.889055</td>\n",
       "      <td>0.183300</td>\n",
       "      <td>1:100000989</td>\n",
       "      <td>A</td>\n",
       "      <td>ATC</td>\n",
       "      <td>1</td>\n",
       "      <td>100000989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          MarkerName Allele1 Allele2  Effect  StdErr  P.value Direction  \\\n",
       "0    1:100000012_G_T       t       g -0.0078  0.0147   0.5950       +-+   \n",
       "1     1:10000006_G_A       a       g  0.0155  0.1109   0.8886       +--   \n",
       "2    1:100000827_C_T       t       c -0.0144  0.0136   0.2915       +-+   \n",
       "3    1:100000843_T_C       t       c  0.0374  0.0289   0.1954       +++   \n",
       "4  1:100000989_A_ATC       a     atc  0.0431  0.0300   0.1499       +++   \n",
       "\n",
       "   HetISq  HetChiSq  HetDf  HetPVal  Pval_IBDseq  Pval_IIBDGC  Pval_GWAS3  \\\n",
       "0    82.7    11.592      2  0.00304     0.239590       0.2746    0.002291   \n",
       "1    54.3     4.373      2  0.11230     0.054628       0.9403    0.404895   \n",
       "2    72.4     7.253      2  0.02661     0.499568       0.6470    0.005517   \n",
       "3     0.0     0.575      2  0.75030     0.387836       0.2238    0.869664   \n",
       "4     0.0     0.835      2  0.65860     0.290844       0.1833    0.889055   \n",
       "\n",
       "   Min_single_cohort_pval       CHR-BP ref  alt CHR         BP  \n",
       "0                0.002291  1:100000012   G    T   1  100000012  \n",
       "1                0.054628   1:10000006   G    A   1   10000006  \n",
       "2                0.005517  1:100000827   C    T   1  100000827  \n",
       "3                0.223800  1:100000843   T    C   1  100000843  \n",
       "4                0.183300  1:100000989   A  ATC   1  100000989  "
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "d1d8f680-c610-4d98-b5f3-4574556fe35c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>MAF</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs201725126</td>\n",
       "      <td>1</td>\n",
       "      <td>13116</td>\n",
       "      <td>0.186879</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs200579949</td>\n",
       "      <td>1</td>\n",
       "      <td>13118</td>\n",
       "      <td>0.186879</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs75454623</td>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>0.479125</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs199856693</td>\n",
       "      <td>1</td>\n",
       "      <td>14933</td>\n",
       "      <td>0.050696</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs78601809</td>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>0.268390</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP  CHR     BP       MAF A1 A2\n",
       "0  rs201725126    1  13116  0.186879  T  G\n",
       "1  rs200579949    1  13118  0.186879  A  G\n",
       "2   rs75454623    1  14930  0.479125  A  G\n",
       "3  rs199856693    1  14933  0.050696  G  A\n",
       "4   rs78601809    1  15211  0.268390  T  G"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "d5b66c83-3c31-46df-86e2-f16c98fd2387",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw = raw.astype({'CHR': int, 'BP': int})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "c2be1961-ed9f-46db-a6cd-bb565e2574a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8404358, 23)"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(raw, ref[['SNP', 'CHR', 'BP', 'A1', 'A2']], on=['CHR', 'BP'], how='inner')\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "027a0799-0efd-4677-b1b0-a46d9a397913",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8404202, 23)"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = merged.loc[((merged.ref==merged.A1) & (merged.alt==merged.A2)) | ((merged.ref==merged.A2) & (merged.alt==merged.A1))]\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "ce2d9e74-2f40-4295-901b-c884d119ce1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P.value</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>...</th>\n",
       "      <th>Pval_GWAS3</th>\n",
       "      <th>Min_single_cohort_pval</th>\n",
       "      <th>CHR-BP</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>922366</th>\n",
       "      <td>2:182308352_A_G</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0924</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>1.218000e-13</td>\n",
       "      <td>---</td>\n",
       "      <td>82.5</td>\n",
       "      <td>11.403</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2.016410e-10</td>\n",
       "      <td>2.016410e-10</td>\n",
       "      <td>2:182308352</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>182308352</td>\n",
       "      <td>rs6740847</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             MarkerName Allele1 Allele2  Effect  StdErr       P.value  \\\n",
       "922366  2:182308352_A_G       a       g -0.0924  0.0125  1.218000e-13   \n",
       "\n",
       "       Direction  HetISq  HetChiSq  HetDf  ...    Pval_GWAS3  \\\n",
       "922366       ---    82.5    11.403      2  ...  2.016410e-10   \n",
       "\n",
       "        Min_single_cohort_pval       CHR-BP  ref  alt CHR         BP  \\\n",
       "922366            2.016410e-10  2:182308352    A    G   2  182308352   \n",
       "\n",
       "              SNP  A1  A2  \n",
       "922366  rs6740847   A   G  \n",
       "\n",
       "[1 rows x 23 columns]"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.loc[merged.SNP=='rs6740847'] # effect on Allele2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "346bd453-3b51-4016-ba87-fe0b662bed18",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.Allele1 = merged.Allele1.str.upper()\n",
    "merged.Allele2 = merged.Allele2.str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "ffb9fe97-eb3c-4711-9b5f-4aec45bff328",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8404202, 23)"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = merged.sort_values(by=['CHR', 'BP']).reset_index(drop=True)\n",
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "5eb05ef7-3a0e-4757-bea9-c73419c869c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[['CHR', 'BP', 'SNP', 'Allele1', 'Allele2', 'Effect', 'StdErr', 'P.value']].to_csv('../blood_coloc/gwas/IBD_2016.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b5dd048-0424-4bd6-983e-0e70ad9cd1a9",
   "metadata": {},
   "source": [
    "## Get index snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "85f545f6-394c-4651-b075-fa51b4f4ff6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8404202, 8)"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw = pd.read_csv('../blood_coloc/gwas/IBD_2016.csv', sep='\\t')\n",
    "raw.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "01b992ae-2d39-49c5-b00a-2390d7767772",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12331, 8)\n"
     ]
    }
   ],
   "source": [
    "pval_thres = 5e-8\n",
    "sig = raw.loc[raw['P.value']<pval_thres].reset_index(drop=True)\n",
    "print(sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "01d65b13-1d67-41fd-8bbc-da635c0d4ae7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P.value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1194804</td>\n",
       "      <td>rs11804831</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0908</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>3.312000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1215424</td>\n",
       "      <td>rs1268339</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0907</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>2.752000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>7918904</td>\n",
       "      <td>rs4908704</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0902</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>3.684000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>7919363</td>\n",
       "      <td>rs17374781</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0898</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>4.487000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>7919825</td>\n",
       "      <td>rs12745962</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0919</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>2.137000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       BP         SNP Allele1 Allele2  Effect  StdErr       P.value\n",
       "0    1  1194804  rs11804831       T       C  0.0908  0.0164  3.312000e-08\n",
       "1    1  1215424   rs1268339       T       C  0.0907  0.0163  2.752000e-08\n",
       "2    1  7918904   rs4908704       A       G -0.0902  0.0164  3.684000e-08\n",
       "3    1  7919363  rs17374781       T       C  0.0898  0.0164  4.487000e-08\n",
       "4    1  7919825  rs12745962       T       C  0.0919  0.0164  2.137000e-08"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "fce90882-4201-4c94-8adf-9272421dc07f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 12331/12331 [00:12<00:00, 1004.20it/s]\n"
     ]
    }
   ],
   "source": [
    "sumstat = sig.copy()\n",
    "sumstat['cyto'] = None\n",
    "for i, row in tqdm.tqdm(sumstat.iterrows(), total=sumstat.shape[0]):\n",
    "    tmp = cytoBand.loc[(cytoBand[0]==f'chr{row.CHR}') & (cytoBand[1]<=row.BP) & (cytoBand[2]>=row.BP)]\n",
    "    sumstat.iloc[i,8] = f'{row.CHR}{tmp[3].values[0]}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "0a1b2c28-96c3-4e47-8d2f-325452758c6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(121, 9)\n"
     ]
    }
   ],
   "source": [
    "index_df = sumstat.loc[sumstat.groupby('cyto')['P.value'].idxmin()].sort_values(by=['CHR', 'BP']).reset_index(drop=True)\n",
    "print(index_df.shape)\n",
    "index_df.to_csv('../blood_coloc/gwas/IBD_2016_indexsnp.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ff6e63-8633-4ad5-b5d9-38ac2651e70f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SpatialScope",
   "language": "python",
   "name": "spatialscope"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
